### PR TITLE
Add icon right-click options (copy icon, icon name, paste icon)

### PR DESCRIPTION
**Pull request by Meorawr**
Until GH support get off their asses and fix Rae's access :meodge:

---

https://github.com/Total-RP/Total-RP-3/assets/172234435/d2d25cf5-37a6-455b-a9fb-26408fdd00c8

Basically adds a right-click dropdown menu to all the locations where the icon browser is available.
> [!NOTE]  
> These options work across your player profile and companions.

https://github.com/Total-RP/Total-RP-3/assets/172234435/29c7f040-9b5c-464a-ac58-f9bcec1a8008

There are three options:
- Copy icon  
Copies the icon from the current frame to an internal clipboard of sorts to be pasted later.
- Copy icon name  
Opens a popup where you can quickly copy the icon name to send it to someone or put it in the icon browser if necessary.
- Paste icon
Pastes the icon you had copied earlier with "Copy icon" to the current icon frame.

> [!TIP]  
> Icon browser locations:
> **Player**
> - Characteristics
>  - Char icon
>  - Misc icons
>  - Trait icons
> - About
>  - Template 2 icons
>  - Template 3 icons
> - Miscellaneous
>  - Glances icons
> 
> **Companion**
> - Information
>  - Char icon
>  - Glances icons
> 
> **Target Frame**
> - Glances bar
>  - Glances icons

> [!IMPORTANT] 
> The following localization was added, but does not require re-translation:
> - `UI_ICON_COPY` (as it took its info from `UI_ICON_BROWSER_HELP`)
> 
> The following localizations were added and require NEW translations:
> - `UI_ICON_OPTIONS`
> - `UI_ICON_COPYNAME`
> - `UI_ICON_PASTE`
> - `REG_COMPANION_ICON` (See reply to this post.)
> - `REG_COMPANION_ICON_TT` (See reply to this post.)
> 
> The following localizations have been removed (they were unused and related):
> - `UI_ICON_BROWSER_HELP`
> - `UI_ICON_BROWSER_HELP_TT`

### CODE REVIEW

**Review by Meorawr — COMMENTED** [review]
I can't explicitly approve but at a first reading this seems fine.

File: totalRP3/Modules/Register/Characters/RegisterAbout.lua [hunk 1]
```diff
@@ -335,11 +344,19 @@ function refreshTemplate2EditDisplay()
...
+		_G[frame:GetName().."Icon"]:SetScript("onMouseDown", function(self, button)
+			if button == "LeftButton" then
+				showIconBrowser(function(icon)
+					frame.frameData.IC = icon;
+					setupIconButton(_G[frame:GetName().."Icon"], icon);
+				end, frameData.IC);
+			elseif button == "RightButton" then
+				TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+					description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(frameData.IC); end);
+					description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({frameData.IC or TRP3_InterfaceIcons.Default}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterAbout.lua:356]: There's a few places where COPY doesn't have a nil check but COPYNAME does. As far as I can tell, these are all places where there always is an icon set by default, so the COPY version makes as much sense as the COPYNAME one, but they should probably be consistent™.

File: totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua [hunk 2]
```diff
@@ -1110,14 +1126,23 @@ function setEditDisplay()
...
+		_G[frame:GetName() .. "Icon"]:SetScript("onMouseDown", function(self, button)
+			if button == "LeftButton" then
+				showIconBrowser(function(icon)
+					miscStructure.IC = icon;
+					setupIconButton(_G[frame:GetName() .. "Icon"], icon or TRP3_InterfaceIcons.Default);
+				end, miscStructure.IC);
+			elseif button == "RightButton" then
+				TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+					description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(miscStructure.IC); end);
+					description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({miscStructure.IC or TRP3_InterfaceIcons.Default}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua:1142]: Here

File: totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua [hunk 3]
```diff
@@ -1200,18 +1225,34 @@ function setEditDisplay()
...
+		frame.CustomLeftIcon:SetScript("onMouseDown", function(self, button)
+			if button == "LeftButton" then
+				showIconBrowser(function(icon)
+					psychoStructure.LI = icon;
+					setupIconButton(self, icon or TRP3_InterfaceIcons.Default);
+				end, psychoStructure.LI);
+			elseif button == "RightButton" then
+				TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+					description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(psychoStructure.LI); end);
+					description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({psychoStructure.LI or TRP3_InterfaceIcons.Default}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua:1237]: Here

File: totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua [hunk 4]
```diff
@@ -1200,18 +1225,34 @@ function setEditDisplay()
...
+		frame.CustomRightIcon:SetScript("onMouseDown", function(self, button)
+			if button == "LeftButton" then
+				showIconBrowser(function(icon)
+					psychoStructure.RI = icon;
+					setupIconButton(self, icon or TRP3_InterfaceIcons.Default);
+				end, psychoStructure.RI);
+			elseif button == "RightButton" then
+				TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+					description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(psychoStructure.RI); end);
+					description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({psychoStructure.RI or TRP3_InterfaceIcons.Default}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua:1252]: Here

File: totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua [hunk 5]
```diff
@@ -1525,7 +1566,17 @@ function TRP3_API.register.inits.characteristicsInit()
...
 	-- UI
 	TRP3_RegisterCharact_Edit_MiscAdd:SetScript("OnClick", miscAddDropDown);
-	TRP3_RegisterCharact_Edit_NamePanel_Icon:SetScript("OnClick", function() showIconBrowser(onPlayerIconSelected, draftData.IC) end);
+	TRP3_RegisterCharact_Edit_NamePanel_Icon:SetScript("onMouseDown", function(self, button)
+		if button == "LeftButton" then
+			showIconBrowser(onPlayerIconSelected, draftData.IC);
+		elseif button == "RightButton" then
+			TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+				description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(draftData.IC); end);
+				description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({draftData.IC or TRP3_InterfaceIcons.ProfileDefault}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua:1575]: Here

File: totalRP3/Modules/Register/Main/RegisterGlance.lua [hunk 6]
```diff
@@ -439,12 +447,21 @@ local function openGlanceEditor(slot, slotData, callback, external, arg1, arg2)
...
+			TRP3_API.popup.hideIconBrowser();
+			if self.isExternal then
+				TRP3_API.popup.showPopup(TRP3_API.popup.ICONS, {parent = TRP3_AtFirstGlanceEditor, point = "RIGHT", parentPoint = "LEFT"}, {onIconSelected, nil, nil, TRP3_AtFirstGlanceEditorIcon.icon});
+			else
+				TRP3_API.popup.showPopup(TRP3_API.popup.ICONS, nil, {onIconSelected, nil, nil, TRP3_AtFirstGlanceEditorIcon.icon});
+			end
+		elseif button == "RightButton" then
+			TRP3_MenuUtil.CreateContextMenu(self, function(_, description)
+				description:CreateButton(loc.UI_ICON_COPY, function() TRP3_API.SetLastCopiedIcon(TRP3_AtFirstGlanceEditorIcon.icon); end);
+				description:CreateButton(loc.UI_ICON_COPYNAME, function() TRP3_API.popup.showCopyDropdownPopup({TRP3_AtFirstGlanceEditorIcon.icon or TRP3_InterfaceIcons.Default}); end);
```
Comment (Solanya) [inline totalRP3/Modules/Register/Main/RegisterGlance.lua:462]: And last one

**Pull request closed without merging**